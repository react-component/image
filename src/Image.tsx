import * as React from 'react';
import { useState } from 'react';
import cn from 'classnames';
import { getOffset } from 'rc-util/lib/Dom/css';
import Preview from './Preview';

export interface ImageProps
  extends Omit<React.ImgHTMLAttributes<HTMLImageElement>, 'placeholder' | 'onClick'> {
  // Original
  src?: string;

  prefixCls?: string;
  previewPrefixCls?: string;
  placeholder?: React.ReactNode;
  fallback?: string;
  preview?: boolean;
  onPreviewClose?: (e: React.SyntheticEvent<HTMLDivElement | HTMLLIElement>) => void;
  onClick?: (e: React.MouseEvent<HTMLDivElement>) => void;
}

type ImageStatus = 'normal' | 'error' | 'loading';

const ImageInternal: React.FC<ImageProps> = ({
  src,
  alt,
  onPreviewClose: onInitialPreviewClose,
  prefixCls = 'rc-image',
  previewPrefixCls = `${prefixCls}-preview`,
  placeholder,
  fallback,
  width,
  height,
  style,
  preview = true,
  className: originalClassName,
  onClick,

  // Img
  crossOrigin,
  decoding,
  loading,
  referrerPolicy,
  sizes,
  srcSet,
  useMap,
  ...otherProps
}) => {
  const isCustomPlaceholder = placeholder && placeholder !== true;
  const [isShowPreview, setShowPreview] = useState(false);
  const [status, setStatus] = useState<ImageStatus>(isCustomPlaceholder ? 'loading' : 'normal');
  const [mousePosition, setMousePosition] = useState<null | { x: number; y: number }>(null);
  const isError = status === 'error';

  const onLoad = () => {
    setStatus('normal');
  };

  const onError = () => {
    setStatus('error');
  };

  const onPreview: React.MouseEventHandler<HTMLDivElement> = e => {
    const { left, top } = getOffset(e.target);

    setShowPreview(true);
    setMousePosition({
      x: left,
      y: top,
    });

    if (onClick) onClick(e);
  };

  const onPreviewClose = (e: React.SyntheticEvent<HTMLDivElement>) => {
    e.stopPropagation();
    setShowPreview(false);
    setMousePosition(null);

    if (onInitialPreviewClose) onInitialPreviewClose(e);
  };

  React.useEffect(() => {
    if (isCustomPlaceholder) {
      setStatus('loading');
    }
  }, [src]);

  const className = cn(prefixCls, originalClassName, {
    [`${prefixCls}-error`]: isError,
  });

  const mergedSrc = isError && fallback ? fallback : src;
  const imgCommonProps = {
    crossOrigin,
    decoding,
    loading,
    referrerPolicy,
    sizes,
    srcSet,
    useMap,
    alt,
    className: cn(`${prefixCls}-img`, {
      [`${prefixCls}-img-placeholder`]: placeholder === true,
    }),
  };

  return (
    <>
      <div
        {...otherProps}
        className={className}
        onClick={preview && !isError ? onPreview : onClick}
        style={{
          ...style,
          width,
          height,
        }}
      >
        {isError && fallback ? (
          <img {...imgCommonProps} src={fallback} />
        ) : (
          <img {...imgCommonProps} onLoad={onLoad} onError={onError} src={src} />
        )}

        {status === 'loading' && (
          <div aria-hidden="true" className={`${prefixCls}-placeholder`}>
            {placeholder}
          </div>
        )}
      </div>
      {preview && !isError && (
        <Preview
          aria-hidden={!isShowPreview}
          visible={isShowPreview}
          prefixCls={previewPrefixCls}
          onClose={onPreviewClose}
          mousePosition={mousePosition}
          src={mergedSrc}
          alt={alt}
        />
      )}
    </>
  );
};

ImageInternal.displayName = 'Image';

export default ImageInternal;
