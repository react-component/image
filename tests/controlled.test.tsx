import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import Image from '../src';

describe('Controlled', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });
  it('With previewVisible', () => {
    const wrapper = mount(
      <Image
        src="https://zos.alipayobjects.com/rmsportal/jkjgkEfvpUPVyRjUImniVslZfWPnJuuZ.png"
        preview={{ visible: true }}
      />,
    );

    expect(wrapper.find('.rc-image-preview').get(0)).toBeTruthy();

    act(() => {
      wrapper.setProps({ preview: { visible: false } });
      jest.runAllTimers();
      wrapper.update();
    });

    expect(wrapper.find('.rc-image-preview').get(0)).toBeFalsy();
  });
});
