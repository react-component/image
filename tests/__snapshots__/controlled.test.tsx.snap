// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controlled With previewVisible 1`] = `
<div
  className="rc-image-preview"
  onClick={[Function]}
  role="document"
  style={Object {}}
>
  <div
    aria-hidden="true"
    style={
      Object {
        "height": 0,
        "outline": "none",
        "overflow": "hidden",
        "width": 0,
      }
    }
    tabIndex={0}
  />
  <div
    className="rc-image-preview-content"
  >
    <div
      className="rc-image-preview-body"
    >
      <ul
        className="rc-image-preview-operations"
      >
        <li
          className="rc-image-preview-operations-operation"
          onClick={[Function]}
        >
          <ForwardRef(CloseOutlined)
            className="rc-image-preview-operations-icon"
          />
        </li>
        <li
          className="rc-image-preview-operations-operation"
          onClick={[Function]}
        >
          <ForwardRef(ZoomInOutlined)
            className="rc-image-preview-operations-icon"
          />
        </li>
        <li
          className="rc-image-preview-operations-operation rc-image-preview-operations-operation-disabled"
          onClick={[Function]}
        >
          <ForwardRef(ZoomOutOutlined)
            className="rc-image-preview-operations-icon"
          />
        </li>
        <li
          className="rc-image-preview-operations-operation"
          onClick={[Function]}
        >
          <ForwardRef(RotateRightOutlined)
            className="rc-image-preview-operations-icon"
          />
        </li>
        <li
          className="rc-image-preview-operations-operation"
          onClick={[Function]}
        >
          <ForwardRef(RotateLeftOutlined)
            className="rc-image-preview-operations-icon"
          />
        </li>
      </ul>
      <div
        className="rc-image-preview-img-wrapper"
        style={
          Object {
            "transform": "translate3d(0px, 0px, 0)",
          }
        }
      >
        <img
          className="rc-image-preview-img"
          onMouseDown={[Function]}
          src="https://zos.alipayobjects.com/rmsportal/jkjgkEfvpUPVyRjUImniVslZfWPnJuuZ.png"
          style={
            Object {
              "transform": "scale3d(1, 1, 1) rotate(0deg)",
            }
          }
        />
      </div>
    </div>
  </div>
  <div
    aria-hidden="true"
    style={
      Object {
        "height": 0,
        "outline": "none",
        "overflow": "hidden",
        "width": 0,
      }
    }
    tabIndex={0}
  />
</div>
`;
